GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,258 @@
+    {
+        // In the context of this project, hydrated files are placeholders or modified paths
+        // The total number of hydrated files is this.PlaceholderCount + this.ModifiedPathsCount
+        private readonly GVFSEnlistmentPathData enlistmentPathData;
+
+        public GVFSEnlistmentHealthCalculator(GVFSEnlistmentPathData pathData)
+        {
+            this.enlistmentPathData = pathData;
+        }
+
+        public GVFSEnlistmentHealthData CalculateStatistics(string parentDirectory)
+        {
+            int gitTrackedItemsCount = 0;
+            int placeholderCount = 0;
+            int modifiedPathsCount = 0;
+            Dictionary<string, int> gitTrackedItemsDirectoryTally = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+            Dictionary<string, int> hydratedFilesDirectoryTally = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+            if (!parentDirectory.EndsWith(GVFSConstants.GitPathSeparatorString) && parentDirectory.Length > 0)
+            {
+                parentDirectory += GVFSConstants.GitPathSeparator;
+            }
+
+            if (parentDirectory.StartsWith(GVFSConstants.GitPathSeparatorString))
+            {
+                parentDirectory = parentDirectory.TrimStart(GVFSConstants.GitPathSeparator);
+            }
+
+            gitTrackedItemsCount += this.CategorizePaths(this.enlistmentPathData.GitFolderPaths, gitTrackedItemsDirectoryTally, parentDirectory);
+            gitTrackedItemsCount += this.CategorizePaths(this.enlistmentPathData.GitFilePaths, gitTrackedItemsDirectoryTally, parentDirectory);
+            placeholderCount += this.CategorizePaths(this.enlistmentPathData.PlaceholderFolderPaths, hydratedFilesDirectoryTally, parentDirectory);
+            placeholderCount += this.CategorizePaths(this.enlistmentPathData.PlaceholderFilePaths, hydratedFilesDirectoryTally, parentDirectory);
+            modifiedPathsCount += this.CategorizePaths(this.enlistmentPathData.ModifiedFolderPaths, hydratedFilesDirectoryTally, parentDirectory);
+            modifiedPathsCount += this.CategorizePaths(this.enlistmentPathData.ModifiedFilePaths, hydratedFilesDirectoryTally, parentDirectory);
+
+            Dictionary<string, decimal> mostHydratedDirectories = new Dictionary<string, decimal>();
+
+            // Map directory names to the corresponding health data from gitTrackedFilesDirectoryTally and hydratedFilesDirectoryTally
+            foreach (KeyValuePair<string, int> pair in gitTrackedItemsDirectoryTally)
+            {
+                if (hydratedFilesDirectoryTally.TryGetValue(pair.Key, out int hydratedFiles))
+                {
+                    // In-lining this for now until a better "health" calculation is created
+                    // Another possibility is the ability to pass a function to use for health (might not be applicable)
+                    mostHydratedDirectories.Add(pair.Key, this.CalculateHealthMetric(hydratedFiles, pair.Value));

```

Comment:
Let me see if I am understanding this calculation correctly and run a scenario by you.  This is taking the number of files and folders for a subfolder of the parent directory and taking the number of hydrated files in that directory to calculate a percentage hydrate for that subfolder right?  So if I have the following folders in a directory:
```
A has 1 file hydrated of 1
B has 1 file hydrated of 1
C has 1 file hydrated of 1
D has 1 file hydrated of 1
E has 1 file hydrated of 1
F has 9,000 files hydrated of 10,000
```

Would the output of the command would be A - E 100%?  or would A - E be ~0.01%?  Seems like we should be using the parent directory total count for calculating the percent so the F will be the most hydrated for that parent?

Reply:
For the design I had in mind, I want the percentages to be the percentages of the specified directory, so A-E would be 100%. However, the act of sorting them by percentage only is a placeholder for now until a better health score is calculated. One of the next steps I plan to take is to calculate a health score and use that instead of just hydrated percentage, and in that case the score for F would be worse than A-E so they would be sorted correctly. I just need time to build out that calculation and figure out what would be the most effective way of doing so, and right now I'm working through the reviews and functional tests to try and get an MVP finished up ASAP 👍